feat: Add a system api call to burn cycles (#230)
dfinity/interface-spec: spec/index.md

    ic0.msg_cycles_accept128 : (max_amount_high : i64, max_amount_low: i64, dst : i32)
                           -> ();                                               // U Rt Ry

>     ic0.burn_cycles : (amount_to_burn : i64) -> (amount : i64);               // U T Rt Ry

Reviewer: Do you also plan a 128-bit version?

Reviewer 2: nit: We typically put the verb at the end of a system call name, so maybe this should be `ic0.cycles_burn` instead?

Reviewer 3: I guess we *only* want the 128-bit version, no? I think we consider the 64-bit versions as deprecated.

Author: My reasoning was that no one will need more the max value of 64-bit cycles to burn in one go. But it is small change so I can change it to 128 if everyone agrees.

Author: It is changed.

Reviewer: Maybe `cycles_burn128` to be consistent with the remaining function names in the 128-bit API.

Author: If you feel strongly about that I can change it, but I prefer it this way because we only have one `cycles_burn` API.

Reviewer: I don't have a strong opinion. How about @Dfinity-Bjoern and @dsarlis ?

Reviewer 2: I can see the arguments of both sides -- consistency with other apis but also in this case we have only the 128-bit api.

Given Björn's comment that we can consider the 64-bit versions deprecated and that one day we might only have 128-bit apis but I don't expect we'll ever go back and change their names, we might choose `cycles_burn128` for the sake of consistency.

But I'm also not feeling strongly about this, I think it's ultimately up to you guys as the spec owners.

Reviewer 3: I would suggest consistency and use `cycles_burn128`.

>     ic0.cycles_burn128 : (amount_high : i64, amount_low : i64, dst : i32) -> ();               // U T Rt Ry

Reviewer 4: Is there a reason why this method cannot be invoked everywhere? i.e.

```suggestion
    ic0.cycles_burn128 : (amount_high : i64, amount_low : i64, dst : i32) -> ();               // * s
```

>     ic0.cycles_burn128 : (amount_high : i64, amount_low : i64, dst : i32) -> ();               // U T Rt Ry

Reviewer: ```suggestion
    ic0.cycles_burn128 : (amount_high : i64, amount_low : i64, dst : i32) -> ();               // I G U Ry Rt C T
```

    This does not trap.

-   `ic0.burn_cycles : (amount_to_burn : i64) → (amount : i64)`

    This burns cycles from the canister. It burns as many cycles as possible, up to these constraints:

    It burns no more cycles than `amount_to_burn`.

>     It moves no more cycles than available according to `balance` - `freezing_threshold`, and

Reviewer: ```suggestion
    It burns no more cycles than available according to `balance` - `freezing_threshold`.
```

>     It moves no more cycles than available according to `balance` - `freezing_threshold`, and

Reviewer: `freezing_threshold` is a value in time units how long the canister should remain installed: we can't deduct time duration from cycles; instead, we need to compute the so-called freezing limit in cycles. I'd reformulate the condition to saying that the canister doesn't become frozen after this system API call.

Author: I am not sure that formulation is correct, as far as I understand canister may become frozen if it has `freezing_limith` cycles left. But it will not be deleted, since it will use those resources to pay for memory and other fixed costs.

Reviewer: Made a concrete edit suggestions => closing this thread.

>     It burns no more cycles than the amount obtained by combining `amount_high` and `amount_low`. Cycles are represented by 128-bit values.

Reviewer 5: What does "combining" mean here? When I saw the names of the parameters I thought these amounts were upper and lower bounds, but here it says "no more cycles than", are we adding them up?

>     It burns no more cycles than available according to `balance` - `freezing_threshold`.

Reviewer: ```suggestion
    It burns no more cycles than `balance` - `freezing_limit`, where `freezing_limit` is the amount of idle cycles burned by the canister during its `freezing_threshold`.
```

    This does not trap.

-   `ic0.cycles_burn128 : (amount_high : i64, amount_low : i64, dst : i32) -> ()`

    This burns cycles from the canister. It burns as many cycles as possible, up to these constraints:

    It burns no more cycles than the amount obtained by combining `amount_high` and `amount_low`. Cycles are represented by 128-bit values.

    It burns no more cycles than `balance` - `freezing_limit`, where `freezing_limit` is the amount of idle cycles burned by the canister during its `freezing_threshold`.

    It can be called multiple times, each time possibly burning more cycles from the balance.

    This call also copies the amount of cycles that were actually burned starting at the location `dst` in the canister memory.

>     This system call does not trap.

Reviewer 6: What if dst is too close to end of memory. Does it truncate the write or trap?

We should specify that it writes to addresses [dst, dst+16), in little-endian, I guess

      es.balance := es.balance + amount
      copy_cycles_to_canister<es>(dst, amount.to_little_endian_bytes())

    ic0.burn_cycles<es>(amount_to_burn : i64) : i64 =

>       if es.context ∉ {U, T, Rt, Ry} then Trap {amount = 0;}

Reviewer: ```suggestion
      if es.context ∉ {U, T, Rt, Ry} then Trap {cycles_used = es.cycles_used;}
```
In case of a Trap we can't return any value.

>       let amount = min(amount_to_burn, es.balance - es.params.sysenv.freezing_threshold)

Reviewer: ```suggestion
      let amount = min(amount_to_burn, es.balance - freezing_limit(
        es.params.sysenv.compute_allocation,
        es.params.sysenv.memory_allocation,
        es.params.sysenv.freezing_threshold,
        memory_usage_wasm_state(es.wasm_state) + es.params.sysenv.memory_usage_raw_module + es.params.sysenv.memory_usage_canister_history,
        es.params.sysenv.subnet_size,
      ))
```

      es.balance := es.balance + amount
      copy_cycles_to_canister<es>(dst, amount.to_little_endian_bytes())

    ic0.cycles_burn128<es>(amount_high : i64, amount_low : i64, dst : i32) =

>       if es.context ∉ {U, T, Rt, Ry} then Trap {cycles_used = es.cycles_used;}

Reviewer: ```suggestion
      if es.context ∉ {I G U Ry Rt C T} then Trap {cycles_used = es.cycles_used;}
```
